### PR TITLE
Add instance and config metadata to user agent

### PR DESCRIPTION
**Pull request by jamesbornholt**
## Description of change

We'd like to record the same metadata that AWS SDKs gather so we can
better understand how different Mountpoint features are used. This
change adds support for detecting platform and instance metadata and
including it in HTTP User-agents. We follow the SDK template for
serializing this metadata.

To make this cleaner, I moved the instance info logic into the client
crate so that all users can get this kind of user agent. The new
`UserAgent` struct supports addings arbitrary key/value pairs, and we
use that in Mountpoint to record basic configurations.

User agents are always a bit annoying to test, but I manually verified
in a few cases (caching enabled/disabled) that this change was sending
the expected headers.

## Does this change impact existing behavior?

No.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/autoconfigure.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+pub fn network_throughput(instance_info: &InstanceInfo) -> anyhow::Result<f64> {
+    let instance_type = instance_info
+        .instance_type()
+        .map_err(|e| anyhow!("failed to get instance type: {e}"))?;
+    let throughput = get_maximum_network_throughput(instance_type).context("failed to get network throughput")?;
+    Ok(throughput)
+}
+
+fn get_maximum_network_throughput(ec2_instance_type: &str) -> anyhow::Result<f64> {
+    const INSTANCE_THROUGHPUT: &str = "instance_throughput";
```
Comment (passaro) [inline mountpoint-s3/src/autoconfigure.rs:17]: Not for this PR, but should we consider using the recommended configuration from https://github.com/awslabs/aws-c-s3/pull/367?

**Review by dannycjones — APPROVED** [review]
LGTM, nice.

**Review by dannycjones — CHANGES_REQUESTED** [review]
Almost LGTM, one typo I forgot I commented

**Review by dannycjones — APPROVED** [review]

File: mountpoint-s3/src/autoconfigure.rs (hunk 1 above)
Comment (jamesbornholt) [reply mountpoint-s3/src/autoconfigure.rs:17]: Right now that API only supports a few hard-coded instance types: https://github.com/awslabs/aws-c-s3/blob/5fe1c3b116d4b2ac69304af2d2d3031e4b6d6d14/source/s3_platform_info.c#L328-L335

I think probably we should consult it first, and then fall back to our version if we don't get a match from it.

Comment (passaro) [reply mountpoint-s3/src/autoconfigure.rs:17]: I was hoping we could remove our list. Not really, unfortunately.

**Pull request merged**